### PR TITLE
feat: issue Stripe refund when camper is deleted

### PR DESCRIPTION
**Pull request by udyding**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=1a6a18a0dfb24ceeabba62338e200e57)

## Implementation description
* Added call to Stripe API to refund and delete users based on charge ID and a given list of camper IDs
* Partial refunds can be done, depending on the given camper IDs to delete and the amount paid for each camper
* Note that currently the secret key being used is the test key (test transactions), and will need to be updated when the project is deployed

## Steps to test
1. Register a new camper so that they are added to the database, and create a test Payment Intent to the Stripe account. There is no code for this yet but you can try running this code before creating the camper, using the generated charge ID for registration:
```
const paymentIntent = await stripe.paymentIntents.create({
        amount: 2000,
         currency: "cad",
         payment_method_types: ["card"],
       });
       const confirmedPaymentIntent = await stripe.paymentIntents.confirm(
         paymentIntent.id,
         { payment_method: "pm_card_visa" },
       );

  const chargeId = confirmedPaymentIntent.charges.data[0].id; 
```
Note that you can register multiple campers under the same charge ID.
2. Using the charge ID, call the `/cancel` endpoint and pass in the charge ID from above, as well as a list of all the campers you want to refund
3. Check the Stripe dashboard to ensure that the payment was made and then refunded (see screenshots)
4. Note that if you want to do partial refunds, the payment in Stripe will be moved to the Refunds section but with only a partial refund.

There are some cases to look for/test:
* When you pass in an empty list of camper IDs, should throw error
* When you pass in a charge ID that doesn't exist or one that does exist but some/all camper IDs do not have such a charge ID, should throw error
* When the amount that you want to refund is larger than the amount in the original payment (usually this happens if there was an error with registration where the `charges` field in the created campers don't add up to the amount in Stripe
* Ensure that in Stripe, when a partial refund is made, the amount refunded is updated/reflected in the Stripe dashboard

## What should reviewers focus on?
* Ensuring all cases pass and there aren't any extra cases that haven't been covered

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

## Screenshots
New payment
image

Refunded payment
image

### CODE REVIEW

**Review by annechung — APPROVED** [review]
tested and looks good! 🚀

**Review by annechung — APPROVED** [review]
took so long to test this guy but works as expected! just a few small comments and should be good to go

File: backend/typescript/services/implementations/camperService.ts [hunk 1]
```diff
@@ -263,7 +269,7 @@ class CamperService implements ICamperService {
       });
 
       try {
-        existingCamp = await MgCamp.findByIdAndUpdate(
+        existingCamp = await MgCampSession.findByIdAndUpdate(
```
[the remark this answers is not in the document]
Comment (udyding) [reply backend/typescript/services/implementations/camperService.ts:272]: Oh the only thing i'm changing is that the camp session is still being imported as `MgCamp`, so I just changed the name of the variable to be accurate with the new changes

File: backend/typescript/services/implementations/camperService.ts [hunk 2]
```diff
@@ -428,21 +457,35 @@ class CamperService implements ICamperService {
...
+      // refund before db deletion - a camper should not be deleted if the refund doesn't go through
+      // calculate amount to be refunded
+      let refundAmount = 0;
+      campersToBeDeleted.forEach((camper) => {
+        const { charges } = camper;
+        refundAmount +=
+          charges.camp + charges.earlyDropoff + charges.latePickup;
+      });
+
+      await stripe.refunds.create({
```
[the remark this answers is not in the document]
Comment (udyding) [reply backend/typescript/services/implementations/camperService.ts:480]: Yes and this error will be caught by the try-catch block

**Pull request merged**